lib/Optimizer/Transforms/FactorQuantumAlloc.cpp

```diff
@@ -42,21 +43,47 @@ class AllocaPat : public OpRewritePattern<quake::AllocaOp> {
     for (std::size_t i = 0; i < size; ++i)
       newAllocs.emplace_back(rewriter.create<quake::AllocaOp>(loc, refTy));
 
-    // 2. Visit all users and replace them accordingly.
-    for (auto *user : allocOp->getUsers()) {
-      if (auto dealloc = dyn_cast<quake::DeallocOp>(user)) {
-        rewriter.setInsertionPoint(dealloc);
-        auto deloc = dealloc.getLoc();
-        for (std::size_t i = 0; i < size - 1; ++i)
-          rewriter.create<quake::DeallocOp>(deloc, newAllocs[i]);
-        rewriter.replaceOpWithNewOp<quake::DeallocOp>(dealloc,
-                                                      newAllocs[size - 1]);
-        continue;
+    std::function<LogicalResult(Operation *, std::int64_t)> rewriteOpAndUsers =
```

Comment:
```suggestion
    auto rewriteOpAndUsers =
```

Reply:
We normally let the compiler type a lambda expression.

Reply:
> We normally let the compiler type a lambda expression.

Agree, but that won't work in this case because this lambda can call itself inside the lambda, and using `auto` returns this compilation error: `a variable declared with an auto type specifier cannot appear in its own initializer`.